### PR TITLE
Allow vanilla (de)serialization of discriminated unions

### LINKED ISSUE #130: Can't have a `Spec` as a field in another model
As a user I would like to use `Spec` (and presumably `Region`) as a field in another pydantic model e.g.

```python
class Foo(BaseModel):
    spec: Spec
```

This produces the following error:

```shell
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_model_construction.py", line 205, in __new__
    complete_model_class(
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_model_construction.py", line 534, in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/main.py", line 643, in __get_pydantic_core_schema__
    return handler(source)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 512, in generate_schema
    schema = self._generate_schema_inner(obj)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 784, in _generate_schema_inner
    return self._model_schema(obj)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 591, in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 591, in <dictcomp>
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 947, in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 1134, in _common_field_schema
    schema = self._apply_annotations(
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 1890, in _apply_annotations
    schema = get_inner_schema(source_type)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 1869, in inner_handler
    from_property = self._generate_schema_from_property(obj, source_type)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 679, in _generate_schema_from_property
    schema = get_schema(
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/scanspec/core.py", line 234, in __get_pydantic_core_schema__
    return handler(source_type)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 789, in _generate_schema_inner
    return self.match_type(obj)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 875, in match_type
    return self._unknown_type_schema(obj)
  File "/home/vid18871/projects/bluesky/blueapi/venv/lib64/python3.10/site-packages/pydantic/_internal/_generate_schema.py", line 415, in _unknown_type_schema
    raise PydanticSchemaGenerationError(
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'scanspec.specs.Spec'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.8/u/schema-for-unknown-type
```

I also tried this, but got the same error:

```python
@discriminated_union_of_subclasses
class Foo(BaseModel):
    spec: Spec
```

### PR DESCRIPTION
**Pull request by DiamondJoseph**
Fixes #130, #131

### CODE REVIEW

**Comment by DiamondJoseph** [comment]
Current handling requires that subclasses of type annotated with `discriminated_union_of_subclasses` are Pydantic dataclasses, which I will add to the docstring.

Types that want to have their schema updated when a new instance of a class decorated with `discriminated_union_of_subclasses` must also either be Pydantic dataclasses or BaseModels.

The FastAPI schema for the routes into the service does not update when a new subclass is created, nor do any types that refer to types that refer to a discriminated union.

If Pydantic exposed some way of propagating that a schema that is used in the schema of a model has updated, so the schema of the model must also update, then `_TaggedUnion` could remove all handling of references and just rebuild_dataclass(`self._baseclass`)

**Comment by DiamondJoseph** [comment]
I've made an issue for that last part, but I don't think it will gain much traction as I think creating runtime models often goes against the purpose of validation

https://github.com/pydantic/pydantic/issues/10091

File: src/scanspec/core.py [hunk 1]
```diff
@@ -104,134 +102,106 @@ def calculate(self) -> int:
...
-    for k, v in get_type_hints(cls_or_func).items():
+    for v in get_type_hints(cls_or_func).values():
         tagged_union = _tagged_unions.get(get_origin(v) or v, None)
         if tagged_union:
-            tagged_union.add_referrer(cls_or_func, k)
+            tagged_union.add_reference(cls_or_func)
     return cls_or_func
 
 
+_tagged_unions: dict[type, _TaggedUnion] = {}
```
Comment (callumforrester) [inline src/scanspec/core.py:160]: Nit: Do we want to attach this to the class itself?

File: src/scanspec/core.py [hunk 2]
```diff
@@ -104,134 +102,106 @@ def calculate(self) -> int:
...
-    def _set_discriminator(self, cls: type | Callable, field_name: str, field: Any):
-        # Set the field to use the `type` discriminator on deserialize
-        # https://docs.pydantic.dev/2.8/concepts/unions/#discriminated-unions-with-str-discriminators
-        if isclass(cls):
-            assert isinstance(
-                field, FieldInfo
-            ), f"Expected {cls.__name__}.{field_name} to be a Pydantic field, not {field!r}"  # noqa: E501
-            field.discriminator = self._discriminator
+        # The members of the tagged union, i.e. subclasses of the baseclass
+        self._members: list[type] = []
```
Comment (callumforrester) [inline src/scanspec/core.py:169]: Nit: `members` is an overloaded term, maybe call this `self._subclasses` to be explicit?

File: src/scanspec/core.py [hunk 3]
```diff
@@ -104,134 +102,106 @@ def calculate(self) -> int:
...
+    @staticmethod
+    # https://github.com/bluesky/scanspec/issues/133
+    def _rebuild(cls_or_func: type | Callable):
+        if isclass(cls_or_func):
+            if is_pydantic_dataclass(cls_or_func):
+                rebuild_dataclass(cls_or_func, force=True)
+            if issubclass(cls_or_func, BaseModel):
+                cls_or_func.model_rebuild(force=True)
+
+    def schema(self, handler):
```
Comment (callumforrester) [inline src/scanspec/core.py:194]: Nit: 

```suggestion
    def schema(self, handler: GetCoreSchemaHandler) -> CoreSchema:
```

File: src/scanspec/core.py [hunk 4]
```diff
@@ -104,134 +102,106 @@ def calculate(self) -> int:
...
         if tagged_union:
-            tagged_union.add_referrer(cls_or_func, k)
+            tagged_union.add_reference(cls_or_func)
     return cls_or_func
 
 
+_tagged_unions: dict[type, _TaggedUnion] = {}
+
+
 class _TaggedUnion:
```
Comment (callumforrester) [inline src/scanspec/core.py:163]: I think it _should_ be possible to move some of this logic back into the `discriminated_union_of_subclasses` function now it is simpler. That might read better with it all in one place? Regardless that should probably be a future issue.

**Review by callumforrester — APPROVED** [review]
LGTM, minor points only, thanks for this!

File: src/scanspec/core.py [hunk 5]
```diff
@@ -104,134 +102,106 @@ def calculate(self) -> int:
         )
 
     Args:
-        super_cls: The superclass of the union, Expression in the above example
+        cls: The superclass of the union, Expression in the above example
```
Comment (ZohebShaikh) [inline src/scanspec/core.py:105]:
```suggestion
        super_cls: The superclass of the union, Expression in the above example
```

**Review by ZohebShaikh — APPROVED** [review]
LGTM, Thank you Joseph

**Pull request merged**